### PR TITLE
[MSHARED-1407] Track dependency usage by referencing classes

### PR DESCRIPTION
**Pull request by jaredstehler**
This enhances the functionality which shows specific classes in use per dependency, by mapping those uses to classes which are referencing them.

The resulting API addition is:

```java
Set<DependencyUsage> analyzeUsages(URL url, ClassesPatterns excludeClasses)
```

which I plan on adding a PR to `maven-dependency-plugin` to show how we use this feature in practice.

This was originally proposed via this PR https://github.com/apache/maven-dependency-analyzer/pull/20 which was closed due to lack of support for default interface methods prior to JDK8.

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MSHARED) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes. Also be sure having selected the correct component.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MSHARED-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MSHARED-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [x] You have run the integration tests successfully (`mvn -Prun-its clean verify`).

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

File: src/main/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzer.java [hunk 1]
```diff
@@ -25,17 +25,11 @@
 import java.io.File;
 import java.io.IOException;
 import java.net.URL;
+import java.util.*;
```
Comment (slawekjaranowski) [inline src/main/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzer.java:28]: please don't use asterix in imports

Comment (jaredstehler) [reply src/main/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzer.java:28]: whoops, we don't use asterix either, so this was an ide glitch on my end. will update.

File: src/main/java/org/apache/maven/shared/dependency/analyzer/ProjectDependencyAnalysis.java [hunk 2]
```diff
@@ -18,14 +18,8 @@
...
 
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.Iterator;
-import java.util.LinkedHashSet;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
```
Comment (slawekjaranowski) [inline src/main/java/org/apache/maven/shared/dependency/analyzer/ProjectDependencyAnalysis.java:21]: 😄

File: src/main/java/org/apache/maven/shared/dependency/analyzer/asm/DependencyClassFileVisitor.java [hunk 3]
```diff
@@ -26,11 +26,8 @@
 
 import org.apache.maven.shared.dependency.analyzer.ClassFileVisitor;
 import org.apache.maven.shared.dependency.analyzer.ClassesPatterns;
-import org.objectweb.asm.AnnotationVisitor;
-import org.objectweb.asm.ClassReader;
-import org.objectweb.asm.ClassVisitor;
-import org.objectweb.asm.FieldVisitor;
-import org.objectweb.asm.MethodVisitor;
+import org.apache.maven.shared.dependency.analyzer.DependencyUsage;
+import org.objectweb.asm.*;
```
Comment (slawekjaranowski) [inline src/main/java/org/apache/maven/shared/dependency/analyzer/asm/DependencyClassFileVisitor.java:30]: 😄

File: src/test/java/org/apache/maven/shared/dependency/analyzer/asm/DependencyVisitorTest.java [hunk 4]
```diff
@@ -20,13 +20,7 @@
...
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
-import org.objectweb.asm.AnnotationVisitor;
-import org.objectweb.asm.Attribute;
-import org.objectweb.asm.FieldVisitor;
-import org.objectweb.asm.Label;
-import org.objectweb.asm.MethodVisitor;
-import org.objectweb.asm.Opcodes;
-import org.objectweb.asm.Type;
+import org.objectweb.asm.*;
```
Comment (slawekjaranowski) [inline src/test/java/org/apache/maven/shared/dependency/analyzer/asm/DependencyVisitorTest.java:23]: 😄

**Comment by jaredstehler** [comment]
Updated with all inadvertant formatting and import changes reverted.

**Pull request merged**